src/Expression/AbstractExpressionBuilder.php

```diff
@@ -0,0 +1,242 @@
+
+declare(strict_types=1);
+
+namespace Yiisoft\Db\Expression;
+
+use Yiisoft\Db\Command\Param;
+use Yiisoft\Db\Connection\ConnectionInterface;
+use Yiisoft\Db\QueryBuilder\QueryBuilderInterface;
+use Yiisoft\Db\Syntax\AbstractSqlParser;
+
+use function array_merge;
+use function count;
+use function strlen;
+use function substr;
+use function substr_replace;
+
+/**
+ * It's used to build expressions for use in database queries.
+ *
+ * It provides a methods {@see build()} for creating various types of expressions, such as conditions, joins, and
+ * ordering clauses.
+ *
+ * These expressions can be used with the query builder to build complex and customizable database queries
+ * {@see Expression} class.
+ *
+ * @psalm-import-type ParamsType from ConnectionInterface
+ */
+abstract class AbstractExpressionBuilder implements ExpressionBuilderInterface
+{
+    public function __construct(private QueryBuilderInterface $queryBuilder)
+    {
+    }
+
+    /**
+     * Builds an SQL expression from the given expression object.
+     *
+     * This method is called by the query builder to build SQL expressions from {@see ExpressionInterface} objects.
+     *
+     * @param Expression $expression The expression to build.
+     * @param array $params The parameters to be bound to the query.
+     *
+     * @psalm-param ParamsType $params
+     *
+     * @return string SQL expression.
+     */
+    public function build(ExpressionInterface $expression, array &$params = []): string
+    {
+        $sql = $expression->__toString();
+        $expressionParams = $expression->getParams();
+
+        if (empty($expressionParams)) {
+            return $sql;
+        }
+
+        if (isset($expressionParams[0])) {
+            $params = array_merge($params, $expressionParams);
+            return $sql;
+        }
+
+        $nonUniqueReplacements = $this->appendParams($expressionParams, $params);
+        $expressionReplacements = $this->buildParamExpressions($expressionParams, $params);
+
+        $replacements = $this->mergeReplacements($nonUniqueReplacements, $expressionReplacements);
+
+        if (empty($replacements)) {
+            return $sql;
+        }
+
+        return $this->replacePlaceholders($sql, $replacements);
+    }
+
+    /**
+     * Appends parameters to the list of query parameters replacing non-unique parameters with unique ones.
+     *
+     * @param array $expressionParams Parameters to be appended.
+     * @param array $params Parameters to be bound to the query.
+     *
+     * @psalm-param ParamsType $expressionParams
+     * @psalm-param ParamsType $params
+     *
+     * @return string[] Replacements for non-unique parameters.
+     */
+    private function appendParams(array &$expressionParams, array &$params): array
+    {
+        $nonUniqueParams = [];
+
+        /** @var non-empty-string $name */
+        foreach ($expressionParams as $name => $value) {
+            $paramName = $name[0] === ':' ? substr($name, 1) : $name;
```

Comment:
Should we support Unicode in parameter names?

Reply:
Placeholder must match the pattern `[:][a-zA-Z0-9_]+` and parameter name may or may not contain `:` at the first position.